langchain/src/storage/convex.ts

```diff
@@ -0,0 +1,210 @@
+ */
+export class ConvexKVStore<
+  T extends Value,
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel>,
+  IndexName extends VectorIndexNames<NamedTableInfo<DataModel, TableName>>,
+  KeyFieldName extends FieldPaths<NamedTableInfo<DataModel, TableName>>,
+  ValueFieldName extends FieldPaths<NamedTableInfo<DataModel, TableName>>,
+  UpsertMutation extends FunctionReference<
+    "mutation",
+    "internal",
+    { table: string; document: object }
+  >,
+  LookupQuery extends FunctionReference<
+    "query",
+    "internal",
+    { table: string; index: string; keyField: string; key: string },
+    object[]
+  >,
+  DeleteManyMutation extends FunctionReference<
+    "mutation",
+    "internal",
+    { table: string; index: string; keyField: string; key: string }
+  >
+> extends BaseStore<string, T> {
+  lc_namespace = ["langchain", "storage", "convex"];
+
+  private readonly ctx: GenericActionCtx<DataModel>;
+
+  private readonly table: TableName;
+
+  private readonly index: IndexName;
+
+  private readonly keyField: KeyFieldName;
+
+  private readonly valueField: ValueFieldName;
+
+  private readonly upsert: UpsertMutation;
+
+  private readonly lookup: LookupQuery;
+
+  private readonly deleteMany: DeleteManyMutation;
+
+  constructor(
+    config: ConvexKVStoreConfig<
+      DataModel,
+      TableName,
+      IndexName,
+      KeyFieldName,
+      ValueFieldName,
+      UpsertMutation,
+      LookupQuery,
+      DeleteManyMutation
+    >
+  ) {
+    super(config);
+    this.ctx = config.ctx;
+    this.table = config.table ?? ("cache" as TableName);
+    this.index = config.index ?? ("byKey" as IndexName);
+    this.keyField = config.keyField ?? ("key" as KeyFieldName);
+    this.valueField = config.valueField ?? ("value" as ValueFieldName);
+    this.upsert =
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      config.upsert ?? (makeFunctionReference("langchain/db:upsert") as any);
+    this.lookup =
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      config.lookup ?? (makeFunctionReference("langchain/db:lookup") as any);
+    this.deleteMany =
+      config.deleteMany ??
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (makeFunctionReference("langchain/db:deleteMany") as any);
+  }
+
+  /**
+   * Gets multiple keys from the Convex database.
+   * @param keys Array of keys to be retrieved.
+   * @returns An array of retrieved values.
+   */
+  async mget(keys: string[]) {
+    return (await Promise.all(
+      keys.map(async (key) => {
+        const found = (await this.ctx.runQuery(this.lookup, {
+          table: this.table,
+          index: this.index,
+          keyField: this.keyField,
+          key,
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } as any)) as any;
+        return found.length > 0 ? found[0][this.valueField] : undefined;
+      })
+    )) as (T | undefined)[];
```

Comment:
I am not sure what `Value` is that `T` extends but is this actually safe? Or does Convex always return e.g. a string?

Reply:
Yup, this is safe. Value is anything Convex serialization supports (strings, numbers, booleans, nulls, arrays, objects, array buffers and int64 big ints): https://docs.convex.dev/functions/args-validation#convex-values